### PR TITLE
Add `ucxx::MemoryHandle` support for multiple memory types

### PR DESCRIPTION
**Pull request by pentschev**
The missing support for specifying memory type to `ucxx::MemoryHandle` prevents usage of non-host memory allocations. With this change it is now possible to create `ucxx::MemoryHandle` with CUDA memory allocation (UCX-managed), as well as pointing the handle to existing CUDA memory allocations.

### CODE REVIEW

**Review by jameslamb — APPROVED** [review]
Approving this as `packaging-codeowners` was added and the changes to `ci/test_common.sh` look fine to me.

But I don't know UCX/ucxx well enough to thoughtfully review the other changes.

File: cpp/include/ucxx/memory_handle.h [hunk 1]
```diff
@@ -28,6 +28,7 @@ class MemoryHandle : public Component {
   ucp_mem_h _handle{};       ///< The UCP handle to the memory allocation.
   size_t _size{0};           ///< The actual allocation size.
   uint64_t _baseAddress{0};  ///< The allocation's base address.
+  ucs_memory_type_t _memoryType{UCS_MEMORY_TYPE_HOST};
```
Comment (madsbk) [inline cpp/include/ucxx/memory_handle.h:31]: Maybe add a comment, now that the other three have one :)

File: cpp/include/ucxx/memory_handle.h [hunk 2]
```diff
@@ -43,12 +44,16 @@ class MemoryHandle : public Component {
...
    *
-   * @param[in] context parent context where to map memory.
-   * @param[in] size    the minimum size of the memory allocation
-   * @param[in] buffer  the pointer to an existing allocation or `nullptr` to allocate a
-   *                    new memory region.
+   * @param[in] context     parent context where to map memory.
+   * @param[in] size        the minimum size of the memory allocation
+   * @param[in] buffer      the pointer to an existing allocation or `nullptr` to allocate a
+   *                        new memory region.
+   * @param[in] memoryType  the type of memory the memory allocation the handle points to.
```
Comment (madsbk) [inline cpp/include/ucxx/memory_handle.h:51]:
```suggestion
   * @param[in] memoryType  the type of memory the handle points to.
```

File: cpp/include/ucxx/memory_handle.h [hunk 3]
```diff
@@ -95,16 +100,18 @@ class MemoryHandle : public Component {
...
```
Comment (madsbk) [inline cpp/include/ucxx/memory_handle.h:107]:
```suggestion
   * @param[in] memoryType  the type of memory the handle points to.
```

File: cpp/include/ucxx/context.h [hunk 4]
```diff
@@ -210,13 +210,15 @@ class Context : public Component {
    *
    * @throws ucxx::Error if either `ucp_mem_map` or `ucp_mem_query` fail.
    *
-   * @param[in] size    the minimum size of the memory allocation
-   * @param[in] buffer  the pointer to an existing allocation or `nullptr` to allocate a
-   *                    new memory region.
+   * @param[in] size        the minimum size of the memory allocation.
+   * @param[in] buffer      the pointer to an existing allocation or `nullptr` to allocate a
+   *                        new memory region.
+   * @param[in] memoryType  the type of memory the memory allocation the handle points to.
```
Comment (madsbk) [inline cpp/include/ucxx/context.h:216]:
```suggestion
   * @param[in] memoryType  the type of memory the handle points to.
```

File: cpp/src/memory_handle.cpp [hunk 5]
```diff
@@ -16,54 +16,65 @@
...
 }
 
 std::shared_ptr<MemoryHandle> createMemoryHandle(std::shared_ptr<Context> context,
                                                  const size_t size,
-                                                 void* buffer)
+                                                 void* buffer,
+                                                 const ucs_memory_type_t memoryType)
 {
-  return std::shared_ptr<MemoryHandle>(new MemoryHandle(context, size, buffer));
+  return std::shared_ptr<MemoryHandle>(new MemoryHandle(context, size, buffer, memoryType));
```
Comment (madsbk) [inline cpp/src/memory_handle.cpp:77]:
```suggestion
  return std::make_shared<MemoryHandle>(context, size, buffer, memoryType);
```

**Review by madsbk — APPROVED** [review]
Looks good, only nitpick

File: cpp/include/ucxx/memory_handle.h (hunk 1 above)
Comment (pentschev) [reply cpp/include/ucxx/memory_handle.h]: Nice catch, done via https://github.com/rapidsai/ucxx/pull/246/commits/a0c0319937780af7dcbb76830f22ea048bb58d5f .

File: cpp/src/memory_handle.cpp (hunk 5 above)
Comment (pentschev) [reply cpp/src/memory_handle.cpp]: Unfortunately here we must use `new`, as the constructors are `private` on purpose and `std::make_shared` doesn't have access to `private` members. 🙁

**Review by pentschev — COMMENTED** [review]
Thanks @madsbk for reviewing, I've applied or responded to the suggested changes.

File: cpp/src/memory_handle.cpp (hunk 5 above)
Comment (pentschev) [reply cpp/src/memory_handle.cpp:77]: Ended up committing this by accident, had to revert it in https://github.com/rapidsai/ucxx/pull/246/commits/d4ad0daf6787f7d2712c15e06caa70cfc5674788 .

**Comment by pentschev** [comment]
Thanks @madsbk and @jameslamb for reviews/approvals. I've also confirmed with the MRC team that this addresses the missing feature they needed so I'm going to go ahead and merge this.

**Pull request merged**